flocker/node/test/test_deploy.py

```diff
@@ -2112,6 +2700,71 @@ def test_volume_resized_before_move(self):
             )])
         self.assertEqual(expected, changes)
 
+    def test_metadata_does_not_cause_restarts(self):
+        """
+        ``P2PNodeDeployer.calculate_necessary_state_changes`` indicates no
+        action necessary if the configuration has metadata for a dataset
+        that is a volume.
+
+        Current cluster state lacks metadata, so we want to verify no
+        erroneous restarts are suggested.
+        """
+        volume_service = create_volume_service(self)
+        node_path = self.successResultOf(volume_service.create(
+            volume_service.get(
+                _to_volume_name(DATASET.dataset_id))
+            )
+        ).get_filesystem().get_path()
+
+        unit = Unit(
+            name=APPLICATION_WITH_VOLUME_NAME,
+            container_name=APPLICATION_WITH_VOLUME_NAME,
+            container_image=APPLICATION_WITH_VOLUME_IMAGE,
+            volumes=frozenset([DockerVolume(
+                container_path=APPLICATION_WITH_VOLUME_MOUNTPOINT,
+                node_path=node_path)]),
+            activation_state=u'active'
+        )
+        docker = FakeDockerClient(units={unit.name: unit})
+
+        current_nodes = [
+            NodeState(
+                hostname=u"node1.example.com",
+                applications={APPLICATION_WITH_VOLUME},
+                manifestations={MANIFESTATION.dataset_id: MANIFESTATION},
+            ),
+        ]
+        manifestation_with_metadata = MANIFESTATION.transform(
+            ["dataset", "metadata"], {u"xyz": u"u"})
+
+        desired_nodes = [
+            Node(
+                hostname=u"node1.example.com",
+                applications={APPLICATION_WITH_VOLUME.transform(
+                    ["volume", "manifestation"], manifestation_with_metadata)},
+                manifestations={MANIFESTATION.dataset_id:
+                                manifestation_with_metadata},
+            ),
+        ]
+
+        # The discovered current configuration of the cluster reveals the
+        # application is running here.
+        current = DeploymentState(nodes=current_nodes)
+        desired = Deployment(nodes=desired_nodes)
+
+        api = P2PNodeDeployer(
+            u"node1.example.com",
+            volume_service, docker_client=docker,
+            network=make_memory_network()
+        )
+
+        changes = api.calculate_necessary_state_changes(
+            self.successResultOf(api.discover_local_state(current_nodes[0])),
+            desired_configuration=desired,
+            current_cluster_state=current,
+        )
+        self.assertEqual(changes, Sequentially(changes=[]))
```

Comment:
:question:

I'm tempted to suggest that a better testing strategy here would be to define the cluster configuration, discover state, calculate changes, **run** the changes, discover state, calculate changes again - and then assert that the changes are null.  This has the benefit of running less fake code.  I'm somewhat afraid that it will turn out to have the major downside of being broken in some random irrelevant way, though.  Try it if you want, I guess.